### ISSUE TITLE
Children in slot not rerendering

### ISSUE THREAD

**Issue by blocka** [vuejs/vue#3386]
### Vue.js version
^2.0.0-beta.5

### Reproduction Link
https://jsfiddle.net/chrisvfritz/rpmce1v8/

### Steps to reproduce
Have one or more branches of code where the same component is rendered with different children. Subsequent renders of one of the other branches, will not show the new children.

```
return this.inner1 
    	? h(Outer, () => 'Inner1') 
      : h(Outer, () => 'Inner2')
```

Set `inner1` to false

### What is Expected?

The screen will show "Inner2"

### What is actually happening?

The screen continues to show "Inner1"

**Comment by yyx990803** [comment]
Ideally you should depend on the condition in the children thunk itself: https://jsfiddle.net/rpmce1v8/1/

The current behavior can be confusing though, I'll see how we can improve that.

**Comment by blocka** [comment]
In particular use case, I have a HOC component which is rendering a component in a layout component. I did some experimenting, and it seems the issues only happens if the HOC is functional.

Here is the version that works https://jsfiddle.net/gxrx0c3f/ and the version that doesn't https://jsfiddle.net/kxvajwcq/

**Comment by yyx990803** [comment]
Fixed in b5b963f51df1fa3124cd8a9846be5455408bc52a and related commits. **Note in the next release it is no longer necessary (and nor supported) to pass component children as thunks. Just directly pass children vnodes!**
